oci/private/pull.bzl

```diff
@@ -199,28 +203,87 @@ def _oci_pull_impl(rctx):
     manifest, size, digest = downloader.download_manifest(rctx.attr.identifier, "manifest.json")
 
     if manifest["mediaType"] in _SUPPORTED_MEDIA_TYPES["manifest"]:
-        if rctx.attr.platform:
-            fail("{}/{} is a single-architecture image, so attribute 'platforms' should not be set.".format(rctx.attr.registry, rctx.attr.repository))
-
+        # plain image manifest: use the contents.
+        pass
     elif manifest["mediaType"] in _SUPPORTED_MEDIA_TYPES["index"]:
+        # image index manifest: download the image manifest for the target platform.
         if not rctx.attr.platform:
-            fail("{}/{} is a multi-architecture image, so attribute 'platforms' is required.".format(rctx.attr.registry, rctx.attr.repository))
+            fail("{}/{} is a multi-architecture image, so attribute 'platforms' is required.".format(
+                rctx.attr.registry,
+                rctx.attr.repository,
+            ))
 
         matching_manifest = _find_platform_manifest(manifest, rctx.attr.platform)
         if not matching_manifest:
-            fail("No matching manifest found in image {}/{} for platform {}".format(rctx.attr.registry, rctx.attr.repository, rctx.attr.platform))
-
-        # extra download to get the manifest for the target platform
+            fail("No matching manifest found in image {}/{} for platform {}".format(
+                rctx.attr.registry,
+                rctx.attr.repository,
+                rctx.attr.platform,
+            ))
         manifest, size, digest = downloader.download_manifest(matching_manifest["digest"], "manifest.json")
     else:
         fail("Unrecognized mediaType {} in manifest file".format(manifest["mediaType"]))
-    
-    # symlink manifest.json to blobs with it's digest. 
+
+    # symlink manifest.json to blobs with its digest.
     # it is okay to use symlink here as copy_to_directory will dereference it when creating the TreeArtifact.
     rctx.symlink("manifest.json", _digest_into_blob_path(digest))
 
-    # download the image config
-    downloader.download_blob(manifest["config"]["digest"], _digest_into_blob_path(manifest["config"]["digest"]))
+    config_output_path = _digest_into_blob_path(manifest["config"]["digest"])
+    downloader.download_blob(manifest["config"]["digest"], config_output_path)
+
+    # if the user provided a platform for the image, validate it matches the config as best effort.
+    if rctx.attr.platform:
+        config_bytes = rctx.read(config_output_path)
+        config = json.decode(config_bytes)
+
+        image_os = config.get("os", None)
+        image_architecture = config.get("architecture", None)
+        image_variant = config.get("variant", None)
+
+        attr_os, _, attr_architecture = rctx.attr.platform.partition("/")
```

Comment:
i am sure we are doing this in multiple places. can move this into an util function?

Reply:
I looks like the opposite is done in `image.bzl`:

```python
def _platform_str(os, arch, variant = None):
    parts = [os, arch]
    if variant:
        parts.append(variant)
    return "/".join(parts)
```

...but after a cursory check, I didn't see an existing platform -> triplet conversion.

I did extract this block into a function as per your sensible suggestion though 👍